File: backend-rest-java-spring-boot/src/main/java/io/specmatic/examples/store/service/InMemoryInventoryClient.java
package io.specmatic.examples.store.service;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.springframework.stereotype.Component;

@Component
public class InMemoryInventoryClient implements InventoryClient {
    private final Map<Integer, Integer> inventory = new ConcurrentHashMap<>();

    public InMemoryInventoryClient() {
        inventory.put(10, 10);
        inventory.put(20, 5);
        inventory.put(30, 7);
    }

    @Override
    public void addInventory(int productId, int count) {
        inventory.merge(productId, count, Integer::sum);
    }

    @Override
    public int getInventory(int productId) {
        return inventory.getOrDefault(productId, 0);
    }

    @Override
    public void reduceInventory(int productId, int count) {
        inventory.compute(productId, (id, current) -> Math.max(1, (current == null ? 1 : current) - count));
    }
}

File: backend-grpc-kotlin-ktor/src/main/kotlin/io/specmatic/samples/store/inventory/InMemoryInventoryClient.kt
package io.specmatic.samples.store.inventory

import io.specmatic.samples.store.persistence.StoreRepository

class InMemoryInventoryClient(private val repository: StoreRepository) : InventoryClient {
    override fun addInventory(productId: Int, count: Int) {
        repository.updateInventory(productId, count)
    }

    override fun getInventory(productId: Int): Int =
        repository.findProduct(productId)?.inventory ?: 0

    override fun reduceInventory(productId: Int, count: Int): Boolean {
        val current = getInventory(productId)
        if (current < count) return false
        repository.updateInventory(productId, current - count)
        return true
    }
}
